[internal/aws] Flaky test TestSweep on Windows

### Component(s)

internal/aws

### Describe the issue you're reporting

Test failure on unrelated change https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/6588556890/job/17901091034?pr=27864#step:6:503

```terminal
make[2]: Entering directory 'D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/aws/metrics'
go test -race -timeout 300s -parallel 4 --tags="" ./...
--- FAIL: TestSweep (0.24s)
    metric_calculator_test.go:276: 
        	Error Trace:	D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/aws/metrics/metric_calculator_test.go:276
        	Error:      	"2.2815ms" is not less than or equal to "1ms"
        	Test:       	TestSweep
FAIL
FAIL	github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/metrics	0.591s
FAIL
```

--- a/internal/aws/metrics/metric_calculator.go
+++ b/internal/aws/metrics/metric_calculator.go
@@ -109,19 +109,48 @@ type MetricValue struct {
 	Timestamp time.Time
 }
 
+// Ticker allows us to mock time.Ticker in unit tests to have more deterministic tests.
+type Ticker interface {
+	C() <-chan time.Time
+	Stop()
+}
+
+type realTicker struct {
+	ticker *time.Ticker
+}
+
+func newRealTicker(d time.Duration) Ticker {
+	return &realTicker{ticker: time.NewTicker(d)}
+}
+
+func (r *realTicker) C() <-chan time.Time {
+	return r.ticker.C
+}
+
+func (r *realTicker) Stop() {
+	r.ticker.Stop()
+}
+
 // MapWithExpiry act like a map which provides a method to clean up expired entries.
 // MapWithExpiry is not thread safe and locks must be managed by the owner of the Map by the use of Lock() and Unlock()
 type MapWithExpiry struct {
-	lock     *sync.Mutex
-	ttl      time.Duration
-	entries  map[any]*MetricValue
-	doneChan chan struct{}
+	lock      *sync.Mutex
+	ttl       time.Duration
+	entries   map[any]*MetricValue
+	doneChan  chan struct{}
+	newTicker func(d time.Duration) Ticker
 }
 
 // NewMapWithExpiry automatically starts a sweeper to enforce the maps TTL. ShutDown() must be called to ensure that these
 // go routines are properly cleaned up ShutDown() must be called.
 func NewMapWithExpiry(ttl time.Duration) *MapWithExpiry {
-	m := &MapWithExpiry{lock: &sync.Mutex{}, ttl: ttl, entries: make(map[any]*MetricValue), doneChan: make(chan struct{}, 1000)}
+	m := &MapWithExpiry{
+		lock:      &sync.Mutex{},
+		ttl:       ttl,
+		entries:   make(map[any]*MetricValue),
+		doneChan:  make(chan struct{}, 1000),
+		newTicker: newRealTicker,
+	}
 	go m.sweep(m.CleanUp)
 	return m
 }
@@ -136,10 +165,11 @@ func (m *MapWithExpiry) Set(key Key, value MetricValue) {
 }
 
 func (m *MapWithExpiry) sweep(removeFunc func(time2 time.Time)) {
-	ticker := time.NewTicker(m.ttl)
+	ticker := m.newTicker(m.ttl)
+
 	for {
 		select {
-		case currentTime := <-ticker.C:
+		case currentTime := <-ticker.C():
 			m.lock.Lock()
 			removeFunc(currentTime)
 			m.lock.Unlock()

--- a/internal/aws/metrics/metric_calculator_test.go
+++ b/internal/aws/metrics/metric_calculator_test.go
@@ -248,34 +248,51 @@ func TestMapKeyNotEqualOnName(t *testing.T) {
 }
 
 func TestSweep(t *testing.T) {
-	sweepEvent := make(chan time.Time)
+	sweepEvent := make(chan time.Time, 2)
+	fakeTicker, push := newFakeTicker()
 	closed := &atomic.Bool{}
-
 	onSweep := func(now time.Time) {
 		sweepEvent <- now
 	}
 
 	mwe := &MapWithExpiry{
-		ttl:      1 * time.Millisecond,
-		lock:     &sync.Mutex{},
-		doneChan: make(chan struct{}),
+		ttl:       0,
+		lock:      &sync.Mutex{},
+		doneChan:  make(chan struct{}),
+		newTicker: func(time.Duration) Ticker { return fakeTicker },
 	}
 
-	start := time.Now()
 	go func() {
 		mwe.sweep(onSweep)
 		closed.Store(true)
 		close(sweepEvent)
 	}()
 
-	for i := 1; i <= 2; i++ {
-		<-sweepEvent
-		clockTime := time.Since(start)
-		require.False(t, closed.Load())
-		assert.LessOrEqual(t, mwe.ttl*time.Duration(i), clockTime)
-	}
+	push(time.Unix(100, 0))
+	push(time.Unix(200, 0))
+
+	t1 := <-sweepEvent
+	t2 := <-sweepEvent
+	assert.False(t, closed.Load())
+	assert.Equal(t, time.Unix(100, 0), t1)
+	assert.Equal(t, time.Unix(200, 0), t2)
+
 	require.NoError(t, mwe.Shutdown())
 	for range sweepEvent { //nolint:revive
 	}
 	assert.True(t, closed.Load(), "Sweeper did not terminate.")
+}
+
+type fakeTicker struct {
+	ch chan time.Time
+}
+
+func (f *fakeTicker) C() <-chan time.Time { return f.ch }
+func (f *fakeTicker) Stop()               {}
+
+func newFakeTicker() (*fakeTicker, func(time.Time)) {
+	ch := make(chan time.Time, 10)
+	ft := &fakeTicker{ch: ch}
+	push := func(t time.Time) { ch <- t }
+	return ft, push
 }

